clarify the meaning of "application-limited" (#139)
NTAP/rfc8312bis: draft-ietf-tcpm-rfc8312bis.md

{: artwork-align="center" }

The elapsed time *t* in {{eq1}} MUST NOT include periods during

> which *cwnd* has not been updated due application-limited behavior

Reviewer: ```suggestion
which *cwnd* has not been updated due to application-limited behavior
```

## Behavior for Application-Limited Flows {#app-limited}

> A flow is application-limited if it is currently sending

Reviewer 2: "currently sending less than cwnd"- is this too general? A pacer or CPU bottleneck could also cause this to happen. Although, maybe it would also make sense to limit cwnd growth in those cases (since the point is that increasing the cwnd will not increase the send rate)?

Reviewer: Pacing the send rate doesn't mean you are sending less than cwnd (excluding app-limited), it just means that you are sending the cwnd paced within a duration (equal to RTT) instead of sending it all at once. As the cwnd growth is based on received ACKs, and ACKs themselves are dependent on sent data, the cwnd sort of grows somewhat paced too (unless the ACKs are coalesced).

I agree that "currently sending less than cwnd" is too generic and maybe we can add text about pacing etc to clarify this. But the rest of the text about cwnd growth looks fine.

Reviewer 2: I'm talking about particular instants in time when I talk about pacing- we will eventually send the whole window, but at the time when we are making a decision about whether to increase the CWnd, we may have deferred sending some of those packets until later in the round. I'm not allowed to look at the linux code, but it's my rough understanding that normally it goes ahead and posts the full CWnd of packets with appropriate "due times" for actually sending each packet, but that's only one possible implementation. TCP itself could also defer sending with a timer (which I think I heard has also been implemented in linux).

Reviewer 2: I'd be alright with leaving this text as Neal has written it for the time being, which is still vague but better than the existing text.

> less than is allowed by the congestion window.

Reviewer: ```suggestion
less than what is allowed by the congestion window.
```

This can happen if the flow is limited by either the
sender application or the receiver application (via the receiver

> advertised window) instead of the sender's congestion window.

Reviewer: I got a little confused when I first read `instead of the sender's congestion window.`
Should we rephrase it to, `... window) and thus sends less data than what is allowed by the sender's congestion window.`

Reviewer 3: This seems to be the last open review comment. I will merge and publish a new rev once the PR shows no requested changes.

Author: Thanks, Vidhi and Lars. This morning's patch should reflect Vidhi's suggestion now:
  https://github.com/NTAP/rfc8312bis/pull/139/files

> advertised window) instead of the sender's congestion window.

Reviewer 3: ```suggestion
advertised window) and thus sends less data than what is allowed by the sender's congestion window.
```

Reviewer 3: I think the change aboveis what @goelvidhi suggested?
